lib/elasticsearch/document.rb

```diff
@@ -31,12 +42,38 @@ def to_document(record)
             primary_record_id: record.id,
             project_model_relationship_id: field[:related_model_id]
           )
-          related_records = relations.map do |related_record|
-            klass = related_record.related_record_type.constantize
-            klass.find(related_record.related_record.id).send(field[:field])
+          if field[:related_type] == 'string'
+            related_records = relations.map do |related_record|
+              # Converts string to callable class name.
+              klass = related_record.related_record_type.constantize
+              klass.find(related_record.related_record.id).send(field[:field])
+            end
+            related_records = related_records.first 
+            document[model_field] = related_records
           end
-          related_records = related_records.first if field[:related_type] == 'string'
-          document[model_field] = related_records
+          if field[:related_type] == "hash"
+            document[model_field] = relations.map do |related_record|
+              props = {}
+              klass = related_record.related_record_type.constantize
+              related_instance = klass.find(related_record.related_record.id)
+            
+              field[:field].each do |prop|
+                if related_instance.respond_to?(prop)
+                  props[prop] = related_instance.send(prop)
+                else
+                  user_defined_field = UserDefinedFields::UserDefinedField.find_by(uuid: prop)
+                  if user_defined_field
+                    props[user_defined_field.column_name] = related_instance.user_defined[prop]
+                  else
+                    Rails.logger.warn("Property #{prop} not found as a method or user-defined field.")
+                  end
+                end
+              end
+              props[:thumbnail_url] = thumbnail_url(props['provider'], props['embed_id'])
+              props
+            end
+            
```

Comment:
You need add another `end` here to close out the if statement that started on line 54.